packages/theme/src/index.js

```diff
@@ -0,0 +1,170 @@
+// @flow
+import * as React from 'react'
+import { isBrowser } from '@emotion/utils'
+
+let canUseCSSVars =
+  isBrowser &&
+  window.CSS &&
+  window.CSS.supports &&
+  window.CSS.supports('--test', 0)
+
+type ThemeType =
+  | $ReadOnly<{
+      [key: string]: ThemeType
+    }>
+  | string
+  | $ReadOnlyArray<ThemeType>
+
+type Ret<Theme> = {
+  Consumer: React.ComponentType<{
+    children: Theme => React.Node
+  }>,
+  consume: <Props>(
+    render: (props: Props, theme: Theme, ref: React.Ref<*>) => React.Node
+  ) => React.ComponentType<Props>,
+  Provider: React.ComponentType<{
+    theme: Theme,
+    children: React.Node,
+    supportsCSSVariables?: boolean
+  }>,
+  Extender: React.ComponentType<{ children: React.Node }>
+}
+
+function getCSSVarUsageTheme(theme: ThemeType, currentPath: string) {
+  if (typeof theme === 'string') {
+    return `var(--theme-${currentPath})`
+  }
+  if (Array.isArray(theme)) {
+    return theme.map((val, i) =>
+      getCSSVarUsageTheme(val, currentPath + '-' + i)
+    )
+  }
+  let keys = Object.keys(theme)
+  let obj = {}
+  for (let i = 0; i < keys.length; i++) {
+    let key = keys[i]
+    obj[key] = getCSSVarUsageTheme(theme[key], currentPath + '-' + key)
+  }
+
+  return obj
+}
+
+function getInlineStyles(
+  theme: ThemeType,
+  currentPath: string,
+  stylesObj: { [string]: string }
+) {
+  if (typeof theme === 'string') {
+    stylesObj[`--theme-${currentPath}`] = theme
+    return stylesObj
+  }
+  if (Array.isArray(theme)) {
+    for (let i = 0; i < theme.length; i++) {
+      getInlineStyles(theme[i], currentPath + '-' + i, stylesObj)
+    }
+    return stylesObj
+  }
+  let keys = Object.keys(theme)
+  for (let i = 0; i < keys.length; i++) {
+    let key = keys[i]
+    getInlineStyles(theme[key], currentPath + '-' + key, stylesObj)
+  }
+
+  return stylesObj
+}
+
+export let createTheme = <Theme: ThemeType>(
+  defaultTheme: Theme
+): Ret<Theme> => {
+  let RawThemeContext = React.createContext(defaultTheme)
+
+  let Context = React.createContext(defaultTheme)
+  // $FlowFixMe this isn't just to get flow to be quiet, i actually want to fix this because i think flow might be right
+  let cssVarUsageTheme: Theme = getCSSVarUsageTheme(defaultTheme, '')
+
+  type ProviderProps = {
+    theme: Theme,
+    children: React.Node,
+    supportsCSSVariables?: boolean
+  }
+
+  let Provider = (props: ProviderProps) => {
+    if (canUseCSSVars) {
+      return (
+        <RawThemeContext.Provider value={props.theme}>
+          <Context.Provider value={cssVarUsageTheme}>
+            <div style={getInlineStyles(props.theme, '', {})}>
+              {props.children}
+            </div>
+          </Context.Provider>
+        </RawThemeContext.Provider>
+      )
+    }
+    return (
+      <Context.Provider value={props.theme}>{props.children}</Context.Provider>
+    )
+  }
+  if (!isBrowser) {
```

Comment:
I’m planning to add a data attribute that says if css vars are used

Reply:
so basically the user will have to decide between deopting to context or causing a render mismatch on IE11, right? not the end of the world

Reply:
The intention is that users would check the userAgent string(I know it's hacky but it's just checking for IE11) and passing the `supportsCSSVariables` prop with whether the client supports CSS Variables. Checking the userAgent string would be opt-in though, so by default it will use context on the server if the prop isn't provided.

Reply:
> Checking the userAgent string would be opt-in though, so by default it will use context on the server if the prop isn't provided.

then that information should be shared between server and client so the client can use context as well automatically to avoid mismatch

Reply:
Yep, hence the data attribute